language/move-vm/runtime/src/runtime.rs

```diff
@@ -111,58 +113,127 @@ impl VMRuntime {
         data_store.publish_module(&module_id, module)
     }
 
-    // See Session::execute_script for what contracts to follow.
-    pub(crate) fn execute_script(
+    fn deserialize_args(&self, tys: &[Type], args: Vec<Vec<u8>>) -> PartialVMResult<Vec<Value>> {
+        // Check if the number of arguments matches the number of types.
+        // REVIEW: is `TYPE_MISMATCH` the most suitable error code here?
+        if tys.len() != args.len() {
+            return Err(
+                PartialVMError::new(StatusCode::TYPE_MISMATCH).with_message(format!(
+                    "argument length mismatch: expected {} got {}",
+                    tys.len(),
+                    args.len()
+                )),
+            );
+        }
+
+        // Get the layouts and kind information of the parameter types.
+        // This could fail under two circumstances:
+        //   1) The loader is in an invalid state.
+        //   2) The outer environment is trying call a Move function with an unsupported parameter type
+        //      (e.g. a reference).
+        // Both are considered invariant violations currently.
+        let mut layouts = vec![];
+        let mut kinfos = vec![];
+        for ty in tys {
+            match self.loader.type_to_type_layout(ty) {
+                Ok(layout) => layouts.push(layout),
+                Err(err) => {
+                    error!("[VM] failed to get layout from type");
+                    return Err(err);
+                }
+            }
+
+            match self.loader.type_to_kind_info(ty) {
+                Ok(kinfo) => kinfos.push(kinfo),
+                Err(err) => {
+                    error!("[VM] failed to get kind info from type");
+                    return Err(err);
+                }
+            }
+        }
+
+        // Try to deserialize the arguments according to the layouts.
+        let mut vals = vec![];
+        for ((layout, kinfo), arg) in layouts.iter().zip(kinfos.iter()).zip(args.iter()) {
+            match Value::simple_deserialize(arg, kinfo, layout) {
+                Some(val) => vals.push(val),
+                None => {
+                    warn!("[VM] failed to deserialize argument");
+                    return Err(PartialVMError::new(
+                        StatusCode::FAILED_TO_DESERIALIZE_RESOURCE,
```

Comment:
`VALUE_DESERIALIZATION_ERROR` or `UNEXPECTED_DESERIALIZATION_ERROR` feel like better fits here given that the argument isn't a resource?